First cut of RocksDB+circlci+cmake on Linux and Mac.

vrdhn opened the pull request:
etl3
====
1. Adding sm-rocksdb:sm-rocksdbjni = 7.7.3-SNAPSHOT to version.props & write-locks.
2. Adding implementation("sm-rocksdb:sm-rocksdbjni") to streaming-code/build.gradle.kts

version
=======
* if commit is tagged with `sm-v7.7.3`, a 7.7.3 version release is pushed to jfrog.
* Else the version is retrieved from version.h, and -SNAPSHOT version is pushed to jfrog.

source
======
* Any addition/removal of java src or test file will have to be updated in java/CMakeLists.txt
* Any addition/removal of test class will have to be updated in class SMRocksJunitRunner

circleci
========
 { build-linux, build-macos, build-windows} contribute jar, test-jar, and jni libraries.
 { make-jar } merges the jni librariesin the jar.
 { test-linux, test-macos, test-windows} run the test-jar on repsective platforms,
 { upload } uploads the merged-jar to artifactory.

TODO
====
1. Add other compressions.
2. mac M1 silicon

vrdhn committed:
First cut of RocksDB+circlci+cmake on Linux, Mac(x86_64,arm64), Windows

etl3
====
1. Adding sm-rocksdb:sm-rocksdbjni = 7.7.3-SNAPSHOT to version.props & write-locks.
2. Adding implementation("sm-rocksdb:sm-rocksdbjni") to streaming-code/build.gradle.kts

version
=======
* if commit is tagged with `sm-v7.7.3`, a 7.7.3 version release is pushed to jfrog.
* Else the version is retrieved from version.h, and -SNAPSHOT version is pushed to jfrog.

source
======
* Any addition/removal of java src or test file will have to be updated in java/CMakeLists.txt
* Any addition/removal of test class will have to be updated in class SMRocksJunitRunner

circleci
========
 { build-linux, build-macos, build-windows} contribute jar, test-jar, and jni libraries.
 { make-jar } merges the jni librariesin the jar.
 { test-linux, test-macos, test-windows} run the test-jar on repsective platforms,
 { upload } uploads the merged-jar to artifactory.

pending
=======
1. Add other compressions.

vrdhn commented on .circleci/config.yml:
```diff
@@ -0,0 +1,256 @@
+      shell: bash.exe
+
+
+## Creates following file:
```
old comment, to be removed.

vrdhn commented on .circleci/config.yml:
```diff
@@ -0,0 +1,256 @@
+      - checkout
+      - run: |
+          set -xe
+          ./build_tools/install_dependencies.sh ../deps jni uring lz4
```
should check that java is 1.8

vrdhn replied:
should check that java is 1.8

vrdhn commented on .circleci/config.yml:
```diff
@@ -0,0 +1,256 @@
+          mkdir release
+          cd release
+          cp /tmp/workspace/jars/rocksdbjni_classes.jar $JAR
+          $JAVA_HOME/bin/jar vuf $JAR -C /tmp/workspace/jars/ librocksdbjni-linux64.so
```
ensure that JAVA_HOME is set and at v1.8

vrdhn commented on .circleci/config.yml:
```diff
@@ -0,0 +1,256 @@
+          xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+          <modelVersion>4.0.0</modelVersion>
+          <groupId>sm-rocksdb</groupId>
+          <artifactId>sm-rocksdbjni</artifactId>
```
should use $GROUPID/$ARTIFACTID here.

vrdhn replied:
should use $GROUPID/$ARTIFACTID here.

vrdhn commented on .circleci/config.yml:
```diff
@@ -0,0 +1,256 @@
+          /usr/bin/find /tmp/workspace
+          ## junit includes hamrest so need some deterministic jar order
+          CP=$(find /tmp/workspace/test-jars /tmp/workspace/release -name '*.jar' | sort  | tr '\n' ':')
+          $JAVA_HOME/bin/java -ea -Xcheck:jni -cp $CP org.rocksdb.test.SMRocksJunitRunner
```
verify that JAVA_HOME is v18

vrdhn commented on .circleci/config.yml:
```diff
@@ -0,0 +1,256 @@
+      JAVA_HOME: /Library/Java/JavaVirtualMachines/adoptopenjdk-8.jdk/Contents/Home
+    steps:
+      - run: HOMEBREW_NO_AUTO_UPDATE=1 brew install --cask adoptopenjdk/openjdk/adoptopenjdk8
+      - attach_workspace:
```
actually we want to test in v18

vrdhn commented on build_tools/install_dependencies.sh:
```diff
@@ -0,0 +1,124 @@
+## Install static dependencies for compiling static rocksdbjava with cmake.
+## Note that .a are made with -fPIC, as they end up in a .so
+## Also note that this should be enhanced to arc specific defines.
+## generate cmake flags
```
arc => arch

vrdhn commented on build_tools/install_dependencies.sh:
```diff
@@ -0,0 +1,124 @@
+with_jni() # ensure JAVA_HOME is set.
+{
+    if [ -z "$JAVA_HOME" ]
+    then
```
should also ensure that v is 1.8, as that's the rocksdb compilation target.

vrdhn replied:
should also ensure that v is 1.8, as that's the rocksdb compilation target.

vrdhn commented on java/CMakeLists.txt:
```diff
@@ -547,3 +668,20 @@ if(NOT MINGW)
+
+
+
+#list( TRANSFORM JAVA_TEST_CLASSES REPLACE  "src/test/java/([^ ]*).java" "\\1" OUTPUT_VARIABLE testfilebase )
```
a test here will not be used by CI, but can be used for development .

bremac replied:
What's the reason that this is commented out? It looks useful to me.

vrdhn replied:
to make this work, the jars has to created in a similar way as being done in .circlci/config, lot of duplication
To make a better local developer exp, the right way is to make a shell script to compile and run tests, and used it in  .circlci/config as well

bremac commented on build_tools/install_dependencies.sh:
```diff
@@ -0,0 +1,139 @@
+{
+    echo Compiling iouring to $DIR/iouring/ >&2
+    cd $DIR/
+    download liburing-0.7.tar.gz https://github.com/axboe/liburing/archive/refs/tags/liburing-0.7.tar.gz
```
What's with the old version of liburing? I believe we're currently on v2.3.

bremac commented on build_tools/install_dependencies.sh:
```diff
@@ -0,0 +1,139 @@
+    tar xf liburing-0.7.tar.gz
+    cd liburing-liburing-0.7/
+    ./configure --prefix=$DIR/iouring/
+    make $J V=1 -C src ENABLE_SHARED=0 CFLAGS=" -g -fomit-frame-pointer -O2 -fPIC" install
```
Why `-g`?

vrdhn replied:
`-g` enables debug info, and is orthogonal to `-O` ... it's just extra bytes in the obj file, which may be useful in case of stack dump

bremac commented on java/CMakeLists.txt:
```diff
@@ -329,6 +436,13 @@ elseif(${CMAKE_VERSION} VERSION_LESS "3.11.4")
+      SOURCES
+      ${JAVA_ALL_TEST_CLASSES}
+      INCLUDE_JARS ${JAVA_TESTCLASSPATH} rocksdbjni_classes
+  )
```
It seems like this is duplicated and can be moved out of the conditional block.

bremac commented on java/CMakeLists.txt:
```diff
@@ -531,14 +651,15 @@ endif()
 set(ROCKSDBJNI_STATIC_LIB rocksdbjni${ARTIFACT_SUFFIX})
 add_library(${ROCKSDBJNI_STATIC_LIB} ${JNI_NATIVE_SOURCES})
-add_dependencies(${ROCKSDBJNI_STATIC_LIB} rocksdbjni_headers)
+add_dependencies(${ROCKSDBJNI_STATIC_LIB} rocksdbjni_headers rocksdbjni_test_classes)
```
Might need a comment here — I can't figure out why the JNI library is dependent on the test JAR.

bremac commented on java/CMakeLists.txt:
```diff
@@ -547,3 +668,20 @@ if(NOT MINGW)
   )
 endif()
+
+if(WITH_LZ4)
+#  find_package(lz4 REQUIRED)
+#  add_definitions(-DLZ4)
+  include_directories(${lz4_INCLUDE_DIRS})
+#  list(APPEND THIRDPARTY_LIBS lz4::lz4)
+endif()
```
Isn't this already being handled when we build the base RocksDB library?

bremac replied:
I did a build with this block commented out, and everything built correctly: https://github.com/sightmachine/rocksdb/commit/4c73a5819948ca7f7d1ad78d58cb153066739ca9

I think we can remove this.

bremac commented on java/CMakeLists.txt:
```diff
@@ -531,14 +651,15 @@ endif()
-  target_link_libraries(${ROCKSDBJNI_SHARED_LIB} ${ROCKSDB_STATIC_LIB} ${ROCKSDB_LIB})
+  add_dependencies(${ROCKSDBJNI_SHARED_LIB} rocksdbjni_headers rocksdbjni_test_classes)
+  target_link_libraries(${ROCKSDBJNI_SHARED_LIB} ${ROCKSDB_STATIC_LIB}
+     ${THIRDPARTY_LIBS} ${SYSTEM_LIBS})
```
Haven't these libraries already been linked to the base RocksDB library?

bremac replied:
I poked around cmake a bit, and discovered that this is because `${ROCKDB_STATIC_LIB}`'s target_link_libraries are set to PRIVATE instead of PUBLIC. As a result, dependent libraries fail to link transitive dependencies. I will push a change to fix this.

bremac replied:
https://github.com/sightmachine/rocksdb/pull/2

bremac commented on java/src/test/java/org/rocksdb/test/SMRocksJunitRunner.java:
```diff
@@ -0,0 +1,106 @@
+
+public class SMRocksJunitRunner {
+
+  // This list has been copied from Makefile
```
Erf. I could see listing the test classes once, but twice is a bit much. In the interests of making this upstreamable, I can throw together a little annotation processor that finds all classes containing methods with an `@Test` annotation.

vrdhn replied:
The test classes are listed in `Makefile`, but we are not using it, so the are relisted here.
The `RocksJunitRunner` can be enhanced with a `-all` option, and  don't even need the SM variant.
However between explicit-list vs detected-list , i really have no preference.

bremac commented:
For parity with the Makefile, I think we should pass `-DCMAKE_INTERPROCEDURAL_OPTIMIZATION=TRUE` to enable LTO.

bremac committed:
Upload artifacts so that we can test them on M1 macs

bremac commented:
It would be helpful to add explicit names to the CircleCI build steps. At the moment, the CircleCI interface is a bit unwieldy since it's displaying the full text of each command where the name should be. :-|

bremac commented:
Artifacts were uploaded to https://app.circleci.com/pipelines/github/sightmachine/rocksdb/86/workflows/af3beb61-3f99-4276-8b96-93d4e2d11de6/jobs/413/artifacts. Will sync up with @deeTEEcee tomorrow to see if he can help run the tests on his M1.

bremac commented:
I was sucked into the IoT Hub vortex yesterday, and today David was out, so we don't have Mac M1 test results yet. However, while I was packaging the artifacts to run on Mac, I tried running them myself and got the following error:

```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGILL (0x4) at pc=0x00007f70221860ed, pid=24462, tid=24463
#
# JRE version: OpenJDK Runtime Environment (19.0.1+10) (build 19.0.1+10)
# Java VM: OpenJDK 64-Bit Server VM (19.0.1+10, mixed mode, tiered, compressed oops, compressed class ptrs, g1 gc, linux-amd64)
# Problematic frame:
# C  [librocksdbjni18181626289572332089.so+0x2f30ed]  ...
#
# An error report file with more information is saved as:
# /home/bremac/tmp/rocksdb-artifacts/hs_err_pid24462.log
#
# ...
#
./run-tests.sh: line 4: 24462 Aborted                 (core dumped) java -ea -Xcheck:jni -cp "$CP" org.rocksdb.test.SMRocksJunitRunner
```

Checking the [error log](https://github.com/sightmachine/rocksdb/files/9993731/hs_err_pid24462.log), I see that the bad instruction sequence starts with 
```
0x00007f70221860ed:   62 f1 ff 08 78 75 f7 49 8d 7d 20 48 39 ce 48 0f
```
which disassembles to an AVX-512 instruction (`vcvttsd2usi`):
```
$ ndisasm -b64 <(echo -n '62 f1 ff 08 78 75 f7 49 8d 7d 20 48 39 ce 48 0f' | sed 's/\(..\) \|\(..\)$/\\\x\1\2/g' | xargs -0 printf)
00000000  62F1FF087875F7    vcvttsd2usi rsi,qword [rbp-0x48]
00000007  498D7D20          lea rdi,[r13+0x20]
0000000B  4839CE            cmp rsi,rcx
0000000E  48                rex.w
0000000F  0F                db 0x0f
```

Somehow `-march=native`, `-DHAVE_PCLMUL`, and `-DHAVE_SSE42` are sneaking into the build and specializing the output for the build server. :-( 

EDIT: Looks like we're just missing `-DPORTABLE=1` on Linux and Windows. I've pushed a fix.

EDIT 2: Confirming that this fixed the problem.

bremac committed:
Ensure that Linux and Windows builds are portable

Discovered when I tried to run the output JAR on my workstation and
it failed when it hit an AVX-512 instruction.

bremac committed:
Fix warning in Windows build

vrdhn committed:
Trying newer version of liburing, and misc

bremac commented on java/CMakeLists.txt:
```diff
@@ -531,14 +644,20 @@ endif()
-  target_link_libraries(${ROCKSDBJNI_SHARED_LIB} ${ROCKSDB_STATIC_LIB} ${ROCKSDB_LIB})
+  ## Some test classes are also generating native functions, so need test_classes here.
+  ##  grep -rlw native java/src/test/
+  add_dependencies(${ROCKSDBJNI_SHARED_LIB} rocksdbjni_headers rocksdbjni_test_classes)
```
I'm still confused here. :confounded:  Is `rocksdbjni_test_classes` somehow standing in for an anonymous set of generated native headers via `GENERATE_NATIVE_HEADERS`?

bremac replied:
Yep, looks like it was. Here's a PR to make that explicit: https://github.com/sightmachine/rocksdb/pull/3

bremac committed:
Fix transitive dependencies

bremac committed:
Make JNI test header dependency explicit

A few of the unit tests depend on JNI headers, so make sure that we
generate the JNI headers explicitly in the build process.

vrdhn committed:
Adding names to circlci, indenting java file, and a typo

bremac committed:
Rename "rocksdbjni-shared" target to "rocksdbjni"

The Java code expects the default JNI library to be named "librocksdbjni.so"
or the equivalent, not "librocksdbjni-shared.so". This commit renames the build
target so that it's easy to run the build output from the command line or
IntelliJ without renaming the shared library.

(Technically, this removes the ability to build a static JNI library. In
practice, I've never heard of anyone using a static JNI library; while it's
technically possible, it requires rebuilding and relinking the JVM.)

vrdhn committed:
Using junit5 console launcher for automatic test detection

bremac committed:
Put all tests classes in JAVA_TEST_CLASSES

This ensures that all of the test classes are present so that they can
be discovered by the JUnit5 runner.

bremac committed:
Merge pull request #5 from sightmachine/vrdhn-junit5

Using junit5 with auto detect

bremac committed:
Hook up CTest for java tests

This commit enables support for running the rocksdbjni tests using CTest.
This is a bit janky (we need to download the junit5 runner from maven),
but it removes some redundant package-munging and makes it possible to
run the tests without hand-crafting the correct command-line invocation.

bremac commented:
Tested on Mac M1 with David Chen and everything worked correctly. :partying_face: I just want to sneak in one last change to make it easier to test changes locally, then I'll merge: https://github.com/sightmachine/rocksdb/pull/7

vrdhn commented:
I messed up a git push and it fried  this repo and closed all the prs .. but this PR can't be reopened.
I think we should review on #9 now.
